File: repeated_calls/orchestrator/agents/repeated_call_agent.py
"""Pre-built Semantic Kernel agent for determining a product issue cause."""

from semantic_kernel import Kernel
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai import FunctionChoiceBehavior
from semantic_kernel.connectors.ai.open_ai import AzureChatPromptExecutionSettings
from semantic_kernel.functions import KernelArguments
from semantic_kernel.agents import AzureAIAgent,AzureAIAgentSettings, AzureAIAgentThread
from repeated_calls.orchestrator.entities.structured_output import RepeatedCallResult
from azure.identity.aio import DefaultAzureCredential
from azure.ai.agents.models import (
    ResponseFormatJsonSchema,
    ResponseFormatJsonSchemaType,
)

# def get_agent(kernel: Kernel, instructions: str) -> ChatCompletionAgent:
#     """Agent for determining if the call is repeated or not."""
#     # Define temperature and which functions the agent can use
#     settings = AzureChatPromptExecutionSettings(
#         function_choice_behavior=FunctionChoiceBehavior.Auto(
#             auto_invoke=True,
#             filters={"included_plugins": ["CustomerDataPlugin"]},
#         ),
#         temperature=0.0,
#         seed=1337,
#         max_tokens=3000,
#         response_format=RepeatedCallResult,
#     )

#     # Create and configure the agent
#     agent = ChatCompletionAgent(
#         name="RepeatedCallAgent",
#         instructions=instructions,
#         kernel=kernel,
#         arguments=KernelArguments(settings=settings),
#         plugins=None,
#     )

#     return agent
async def get_agent_response(instructions: str, userprompt:str, thread: AzureAIAgentThread) -> str:
    """Agent for determining if the call is repeated or not."""
    # Define temperature and which functions the agent can use
    ai_agent_settings = AzureAIAgentSettings()
    
    async with (
        DefaultAzureCredential() as creds,
        AzureAIAgent.create_client(credential=creds, 
                                    conn_str=ai_agent_settings.endpoint, 
                                    deployment_name=ai_agent_settings.model_deployment_name) as client,
    ):
        # Create agent definition
        agent_definition = await client.agents.create_agent(
            model=ai_agent_settings.model_deployment_name,
            name="RepeatedCallAgent",
            instructions=instructions,
            response_format=ResponseFormatJsonSchemaType(
                json_schema=ResponseFormatJsonSchema(
                    name="repeated_call",
                    description="Extract repeated call information.",
                    schema=RepeatedCallResult.model_json_schema(),
                )
            )
        )

        # Create the AzureAI Agent
        agent = AzureAIAgent(
            client=client,
            definition=agent_definition,
        )

        response = await agent.get_response(
            messages=userprompt,
            thread=thread
        )

    return response.content

File: dotnet-samples/Agents/AzureAI/RepeatedCallAgent.cs
using Azure.AI.Agents.Persistent;
using Azure.Identity;
using cw_repeated_calls_dotnet.Agents.SemanticKernel;
using cw_repeated_calls_dotnet.Entities.States;
using cw_repeated_calls_dotnet.Entities.StructuredOutput;
using cw_repeated_calls_dotnet.PromptEngineering;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Agents;
using Microsoft.SemanticKernel.Agents.AzureAI;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;
using ModelContextProtocol.Protocol;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace cw_repeated_calls_dotnet.Agents.AzureAI
{
    public class RepeatedCallAgent: IAgent<RepeatedCallResult>
    {
        public async Task<RepeatedCallResult> GetAgentResponseAsync(Kernel kernel, RepeatedCallState state)
        {
            var prompts = new RepeatedCallPrompts(state);

            Kernel agentKernel = kernel.Clone();
            var plugins = agentKernel.Plugins;

            var client = agentKernel.GetRequiredService<PersistentAgentsClient>();
            var chatCompletion = kernel.GetRequiredService<IChatCompletionService>();
            string model = chatCompletion.Attributes["DeploymentName"].ToString();

            var definition = await client.Administration.CreateAgentAsync(
               model: model,
               name: "RepeatedCallAgent",
               instructions: prompts.SystemPrompt,
               responseFormat: BinaryData.FromString(JsonSerializer.Serialize(new
               {
                   type = "json_schema",
                   json_schema = new
                   {
                       name = "repeatedcall",
                       description = "Repeated call schema.",
                       schema = new
                       {
                           type = "object",
                           properties = new
                           {
                               customerId = new { type = "integer", description = "The ID of the customer" },
                               analysis = new { type = "string", description = "Analysis of the problem" },
                               conclusion = new { type = "string", description = "Conclusion of the analysis" },
                               isRepeatedCall = new { type = "boolean", description = "Whether this is a repeated call or not" },
                           },
                           required = new[] { "customerId", "analysis", "conclusion", "isRepeatedCall" }
                       }
                   }
               }))
               );

            AzureAIAgent repeatedCallAgent = new(definition, client, plugins: agentKernel.Plugins)
            {

                Arguments = new KernelArguments(
                    new OpenAIPromptExecutionSettings()
                    {
                        Temperature = 0,
                        FunctionChoiceBehavior = FunctionChoiceBehavior.Auto(options: new() { RetainArgumentTypes = true }),
                    })
            };

            AzureAIAgentThread thread = new AzureAIAgentThread(client, state.ThreadId);
            ChatMessageContent message = new(AuthorRole.User, prompts.UserPrompt);

            RepeatedCallResult repeatedCallResult = null;
            try
            {
                ChatMessageContent response = await repeatedCallAgent.InvokeAsync(message, thread: thread).FirstAsync();
                repeatedCallResult = JsonSerializer.Deserialize<RepeatedCallResult>(response.Content!.ToString());
            }
            finally
            {
                await client.Administration.DeleteAgentAsync(repeatedCallAgent.Id);
            }

            return repeatedCallResult ?? throw new InvalidOperationException("Failed to deserialize the response from the Cause Agent.");

        }
    }
}
